src/install/hosted_git_info.zig

```diff
@@ -0,0 +1,601 @@
+        .sourcehut => .{
+            .protocols = &.{ .git_plus_ssh, .https },
+            .domain = "git.sr.ht",
+            .shortcut = "sourcehut:",
+            .tree_path = "tree",
+            .blob_path = "tree",
+            .edit_path = null,
+            .edit_template = 0,
+            .tarball_template = 0,
+            .extract = 0,
+        },
+    };
+}
+
+/// Search for the appropriate `HostInfo` by the protocol string.
+fn findHostInfoByProtocol(protocol: []const u8) ?HostInfo {
+    inline for (@typeInfo(Host).@"enum".fields) |field| {
+        const host: Host = @enumFromInt(field.value);
+        const info = getHostInfo(host);
+        // We slice the last character off since URL.protocol() returns foo out of foo:// and
+        // info.shortcut includes the colon.
+        if (std.mem.eql(u8, info.shortcut[0 .. info.shortcut.len - 1], protocol)) {
+            return info;
+        }
+    }
+    return null;
+}
+
+fn findHostInfoByDomain(hostname: []const u8) ?HostInfo {
+    inline for (@typeInfo(Host).@"enum".fields) |field| {
+        const host: Host = @enumFromInt(field.value);
+        const info = getHostInfo(host);
+        if (std.mem.eql(u8, info.domain, hostname)) {
+            return info;
+        }
+    }
+    return null;
+}
+
+/// Search the the appropriate `HostInfo` by deducing it from the URL.
+fn deduceHostInfo(allocator: std.mem.Allocator, url: *bun.jsc.URL) ?HostInfo {
+    const proto_str = url.protocol();
+    if (findHostInfoByProtocol(proto_str.byteSlice())) |host_info| {
+        return host_info;
+    }
+
+    // TODO(markovejnovic): I don't know if this conversion is correct.
+    const as_slice = url.hostname().toSlice(allocator);
+    defer as_slice.deinit();
+    const hostname = bun.strings.withoutPrefixComptime(as_slice.slice(), "www.");
+    if (findHostInfoByDomain(hostname)) |host_info| {
+        return host_info;
+    }
+
+    return null;
+}
+
+/// Test whether the given node-package-arg string is a GitHub shorthand.
+///
+/// This mirrors the implementation of hosted-git-info, though it is significantly faster.
+fn isGithubShorthand(
+    npa_str: []const u8,
+) bool {
+    // The implementation in hosted-git-info is a multi-pass algorithm. We've opted to implement a
+    // single-pass algorithm for better performance.
+    //
+    // This could be even faster with SIMD but this is probably good enough for now.
+    if (npa_str.len < 1) {
+        return false;
+    }
+
+    // Implements doesNotStartWithDot
+    if (npa_str[0] == '.' or npa_str[0] == '/') {
+        return false;
+    }
+
+    var pound_idx: ?u32 = null;
+    var seen_slash = false;
+
+    for (npa_str, 0..) |c, i| {
+        switch (c) {
+            // Implement atOnlyAfterHash and colonOnlyAfterHash
+            ':', '@' => {
+                if (pound_idx == null) {
+                    return false;
+                }
+            },
+
+            '#' => {
+                pound_idx = @intCast(i);
```

Comment:
Can we just make `pound_idx` a `?usize` so we don't have to cast?

Reply:
I want to do it as a future improvement since time is running out, if that's OK